### PR TITLE
Add ExcisionSphere to Domain/Structure and test

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes
Adds the class ExcisionSphere to Domain/Structure, which holds the computational domain's excision sphere
information, including the center and radius, as well as an ID for when more than one excision surface is present.

Please see Further Comments for more information.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

This is the first in a series of PRs that will place the computational domain's excision
sphere information into the Domain where it can be used to the control system. This PR
adds the ExcisionSphere class, and domains with excision spheres (Shell, BinaryCompactObject) will hold maps of strings to ExcisionSpheres. These will be added in future PRs.

### CODE REVIEW

File: src/Domain/Structure/ExcisionSphere.cpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.cpp:19]: Make the check and message consistent.  (Is 0 allowed?)

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.hpp]: Design question: I don't know exactly how this is intended to be used, but should the ID be stored in this class, or would this class be the value in a map with the ID as a key?  What is the ID used for?

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.hpp:29]: Does this class make sense in VolumeDim != 3?  I don't think it makes sense in VolumeDim = 1, at least.

File: tests/Unit/Domain/Structure/Test_ExcisionSphere.cpp
Comment (wthrowe) [inline tests/Unit/Domain/Structure/Test_ExcisionSphere.cpp]: const stuff.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (mar-celine) [reply src/Domain/Structure/ExcisionSphere.hpp:29]: I considered it could be used for VolumeDim = 2 in the case where we have a Disk with an excised region - in 1D the excised region would be an interval, and we would have a disconnected domain - it makes sense mathematically (i think) but it's not anything I've seen used in SpECTRE

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (mar-celine) [reply src/Domain/Structure/ExcisionSphere.hpp]: @nilsdeppe mentioned using a map from strings to excision spheres, so the string could be the key, and the ID could be the hash of the string (am I remembering/explaining this correctly @nilsdeppe ?)

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (wthrowe) [reply src/Domain/Structure/ExcisionSphere.hpp:29]: I'm not sure disconnected domains are useful, but you're right that 2D could be, so let's leave it general.

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (wthrowe) [reply src/Domain/Structure/ExcisionSphere.hpp]: But what is the ID for?  What are you going to do with it?

Comment (nilsdeppe) [reply src/Domain/Structure/ExcisionSphere.hpp]: That's how I suggested generating the ID, I'm just not sure what we want to use the ID for. I'd expect places that need the ID will just do `domain.excision_spheres().at("ExcisionA")` where `domain.excision_spheres()` returns a `unordered_map<string, ExcisionSphere>`. Having said all that, @wthrowe probably knows the use-cases in the control system better than I do :)

Comment (mar-celine) [reply src/Domain/Structure/ExcisionSphere.hpp]: @wthrowe just to confirm, should I remove the ID to move this PR along? I'm fine with removing it.

Comment (wthrowe) [reply src/Domain/Structure/ExcisionSphere.hpp]: Yes, that's fine.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Go ahead and squash, including the ID stuff.

**Review by nilsdeppe — DISMISSED** [review]

File: tests/Unit/Domain/Structure/Test_ExcisionSphere.cpp
Comment (wthrowe) [inline tests/Unit/Domain/Structure/Test_ExcisionSphere.cpp]: This has to be the first include.

File: src/Domain/Structure/ExcisionSphere.cpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.cpp]: Merge blocks.

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.hpp]: Use `NOLINTNEXTLINE(google-runtime-references)`

**Comment by mar-celine** [comment]
@wthrowe hopefully this PR is ready to be merged! I have pushed the second PR in this series

File: src/Domain/Structure/ExcisionSphere.cpp
Comment (nilsvu) [inline src/Domain/Structure/ExcisionSphere.cpp:16]: Question: How does this class relate to excision surfaces that are not coordinate-spheres? Does it assume a particular frame in which the excision is a coordinate-sphere, in which case should we template it on the frame?

**Review by wthrowe — DISMISSED** [review]
I have a few minor requests, but if you wish you may push them to a short followup PR.  You should also answer @nilsleiffischer's question before this is merged.

File: src/Domain/Structure/ExcisionSphere.cpp
Comment (mar-celine) [reply src/Domain/Structure/ExcisionSphere.cpp:16]: The ExcisionSphere class is written to only hold information for excision surfaces in the grid frame, which are assumed to be coordinate spheres. Information about excision surfaces which are not coordinate spheres, such as spherical harmonic information, will be known by the shape map, so there will be no ExcisionSphere object corresponding to them. So the frame in which the excision is a coordinate sphere is assumed to be the grid frame always, so it is not templated on frame.

But I will add comment(s) about this to the class so that that is explained :slightly_smiling_face:

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.hpp]:
```
// NOLINTNEXTLINE(google-runtime-references)
void pup(PUP::er& p) noexcept;
```

File: src/Domain/Structure/ExcisionSphere.cpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.cpp]: Alphabetize.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Domain/Structure/ExcisionSphere.hpp
Comment (wthrowe) [inline src/Domain/Structure/ExcisionSphere.hpp]: Remove this line.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Domain/Structure/ExcisionSphere.cpp
Comment (nilsvu) [reply src/Domain/Structure/ExcisionSphere.cpp:16]: ok thanks 👍 I would suggest to add pretty much your entire comment here to the docstring, but I'm satisfied with this and there's no need to wait for my approval on this PR.

**Review by nilsdeppe — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**
